capsules/extra/src/process_info_driver.rs

```diff
@@ -0,0 +1,254 @@
+//!   on the board. Returns the number of running processes.
+//! - 3: Fill the allow RW buffer with the short IDs for the running processes
+//!   on the board. Returns the number of running processes.
+//! - 4: Put the name of the process specified by the process ID in `data1` in
+//!   the allow RW buffer.
+//! - 5: Fill the allow RW buffer with the following information for the process
+//!   specified by the process ID in `data1`.
+//!   - The number of timeslice expirations.
+//!   - The number of syscalls called.
+//!   - The number of restarts.
+//!   - The current process state (running=0, yielded=1, yieldedfor=2,
+//!     stopped=3, faulted=4, terminated=5).
+//! - 10: Change the process state. `data1` is the process ID, and `data2 is the
+//!   new state.(1=start, 2=stop, 3=fault, 4=terminate, 5=boot).
+
+use kernel::capabilities::{ProcessManagementCapability, ProcessStartCapability};
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::process;
+use kernel::processbuffer::WriteableProcessBuffer;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::Kernel;
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::ProcessInfo as usize;
+
+mod rw_allow {
+    pub const INFO: usize = 0;
+    /// The number of allow buffers the kernel stores for this grant
+    pub const COUNT: u8 = 1;
+}
+
+pub struct ProcessInfo<C: ProcessManagementCapability> {
+    apps: Grant<(), UpcallCount<0>, AllowRoCount<0>, AllowRwCount<{ rw_allow::COUNT }>>,
+    /// Reference to the kernel object so we can access process state.
+    kernel: &'static Kernel,
+    /// Capability needed to interact with and control processes.
+    capability: C,
+}
+
+impl<C: ProcessManagementCapability> ProcessInfo<C> {
+    pub fn new(
+        kernel: &'static Kernel,
+        grant: Grant<(), UpcallCount<0>, AllowRoCount<0>, AllowRwCount<{ rw_allow::COUNT }>>,
+        capability: C,
+    ) -> Self {
+        Self {
+            kernel,
+            apps: grant,
+            capability,
+        }
+    }
+
+    fn iterate_u32<F>(&self, process_id: ProcessId, func: F) -> u32
+    where
+        F: Fn(&dyn kernel::process::Process) -> u32,
+    {
+        let mut count = 0;
+        let _ = self.apps.enter(process_id, |_app, kernel_data| {
+            let _ = kernel_data
+                .get_readwrite_processbuffer(rw_allow::INFO)
+                .and_then(|shared| {
+                    shared.mut_enter(|s| {
+                        let mut chunks = s.chunks(size_of::<u32>());
+
+                        self.kernel
+                            .process_each_capability(&self.capability, |process| {
+                                // Get the next chunk to write the next
+                                // PID into.
+                                if let Some(chunk) = chunks.nth(0) {
```

Comment:
```suggestion
                                if let Some(chunk) = chunks.next() {
```

Weird use of `.nth()`, `.next()` _should_ be equivalent?

Reply:
Oh. I see what happened. Rust, what is the difference between a "required method" and a "provided method", and why as a developer do I care about the difference? https://doc.rust-lang.org/core/iter/trait.Iterator.html

Reply:
"Required" means an instance of the trait has to implement it themselves. "Provided" means the trait comes with a default implementation (typically based on required methods). As a user of a trait you don't care, but as an implementer of a trait, you definitely care and they are kind of the opposite.